sys-cluster/slurm/files/slurmd.initd

```diff
@@ -0,0 +1,79 @@
+#!/sbin/openrc-run
+# Copyright 1999-2018 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+extra_commands="reload"
+
+[[ -f /etc/conf.d/slurm ]] && . /etc/conf.d/slurm
+
+depend() {
+	local _need
+	[[ ${SLURM_USE_MUNGE} -ne 0 ]] && _need="${_need} munged"
+	[[ ${SLURM_USE_YPBIND} -ne 0 ]] && _need="${_need} ypbind"
+	after logger
+	need ${_need}
+}
+
+create_folder_and_set_permission() {
+    folder=${1}
+    # Fix permissions (/tmp is wiped at boot)
+    if [[ "${folder}" == "/tmp" ]]; then
+        eerror "SLURM: folder == ${folder} == /tmp"
+        exit
+    fi
+    if [[ "${folder}" == "/var/tmp" ]]; then
+        eerror "SLURM: folder == ${folder} == /var/tmp"
+        exit
+    fi
+    checkpath -d -o slurm:slurm ${folder}
+}
+
+checkconfig() {
+	if [ ! -e "/etc/slurm/slurm.conf" ]; then
+		eerror "Missing config /etc/slurm/slurm.conf"
+		eerror "Customize sample one or generate new by configurator.html"
+		eerror "in slurm doc directory"
+		return 1
+	fi
+	# Make sure folders exists
+	. /etc/slurm/slurm.conf
+	folders=(
+		${SlurmdSpoolDir}
+		$(dirname ${SlurmctldLogFile})
+		$(dirname ${SlurmctldPidFile})
+		$(dirname ${SlurmdLogFile})
+		$(dirname ${SlurmdPidFile})
+		${SlurmdSpoolDir}
```

Comment:
`SlurmdSpoolDir` appears twice :)

More importantly, there is a potential for an exploit here. Imagine the following scenario as root:

```
# chown slurm:slurm /foo
# chown slurm:slurm /foo/bar
```

When the two commands are run in that order, there is a small window where the slurm user can replace `/foo/bar` with a link. The `checkpath` utility tries hard to guard against that, but there's still a race condition.

In my opinion it's simpler to pick reasonable default values for these paths, and make them work out-of-the-box by setting permissions in the ebuild. When everything works, users will be much less likely to mess with them, and then you don't have to worry about "fixing" the permissions in the first place. Up to you, but I had to mention it!

Reply:
You are right Michael, the default `slurm.conf.example` gives:

```
SlurmctldPidFile=/run/slurm/slurmctld.pid
SlurmdPidFile=/run/slurm/slurmd.pid
SlurmdSpoolDir=/var/spool/slurm/slurmd
SlurmctldLogFile=/var/log/slurmctld.log
SlurmdLogFile=/var/log/slurmd.log
```

The directories `/var/log` exists, `/var/spool/slurm` is created by the ebuild, and `/run/slurm` is managed by tmpfiles.  If users change the default, I will let the script exit if the needed directories does not exist.

Reply:
Hmm, after reading the manpage of `openrc-run(8)`, `checkpath` seems to guard against the symlink attack, although I don't know the details underneath.   I shall fix the duplication for the moment and keep this security concern in mind.

Reply:
Symlinks should be safe but there's still a race condition in the hardlink detection. In pseudo-code, checkpath does,

```
if this inode has more than one link {
  error
}
else {
  change the permissions
}
```

but it's possible for the number of links to change between the first and last line. On linux these attacks are preventable with the `fs.protected_hardlinks` sysctl so they're not a huge problem. But I've found that refactoring to avoiding them tends to make the code simpler anyway.

Reply:
Interesting!  I read about `protected_hardlinks` and got the same impression.  Noted.

Thanks Michael for the enlightenment.